Sonos setup fails with unhandled exceptions on discovery messages

PeteRager opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
During setup of manually configured Sonos devices, if the devices are offline setup of the integration may fail.  There were several issues in the code that needed to be resolved.

1. Exceptions were not being handled when the device was unable to be pinged and were flowing back to HASS during setup.  Now the exceptions are handled and logged, setup will complete, and the devices will be periodically retried.
2. During initialization, a blocking call to soco.uuid was occurring within the HASS event loop, this issue is mitigated by not attempting the ping if the prior call to get the uuid via async_add_executor_job failed.
3. If an exception occurred when pinging the device, the call to async_call_later (to retry it) would not occur and hence the device would never come back on-line without a HASS restart.
4. In the attached issues asyncio.TimeoutError was unhandled, during my testing I saw RequestException occur.  Added these to the existing exception handler to make them consistent.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #90226
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

PeteRager committed:
Sonos setup fails with unhandled exceptions on discovery messages

jjlawren commented on homeassistant/components/sonos/__init__.py:
```diff
@@ -12,7 +12,7 @@
 
 from aiohttp import ClientError
-from requests.exceptions import Timeout
+from requests.exceptions import RequestException, Timeout
```
```suggestion
from requests.exceptions import Timeout
```
Importing and catching `RequestException` is unnecessary as we're already catching `OSError`. Here you can see that `RequestException` is a subclass of `IOError`, which has been merged into `OSError`:
https://github.com/psf/requests/blob/7f694b79e114c06fac5ec06019cada5a61e5570f/requests/exceptions.py#L12

See version change notes for `OSError` here: https://docs.python.org/3/library/exceptions.html#OSError

PeteRager replied:
Excellent point, thank you.  On unrelated note, I found another issue in this routine when I did extensive system testing - plugging and unplugging speakers - simple fix though it required more unit tests - will have that change and this tomorrow.

PeteRager replied:
Fixed

jjlawren commented on homeassistant/components/sonos/__init__.py:
```diff
@@ -353,7 +353,13 @@ def get_sync_attributes(soco: SoCo) -> set[SoCo]:
+                SoCoException,
+                Timeout,
+                asyncio.TimeoutError,
+                RequestException,
```
```suggestion
```

PeteRager replied:
fixed

jjlawren commented on homeassistant/components/sonos/__init__.py:
```diff
@@ -383,30 +388,46 @@ def get_sync_attributes(soco: SoCo) -> set[SoCo]:
+                        SoCoException,
+                        Timeout,
+                        asyncio.TimeoutError,
+                        RequestException,
```
```suggestion
```

PeteRager replied:
fixed

PeteRager committed:
Fix event loop blocking call, speakers not going unavailable

1. Speaker activity messages were being sent for speakers whose pings fail, resulting in the speaker remaining available in HASS.  Now the activity message is only sent after a successful ping.
2.  Remove RequestException from except handler as its a derived class of OSError()
3. If the second host in a list of two hosts was unavailable, then a blocking call would occur in the event loop.  Now, fully process the lists of hosts in the first loop; and only process hosts there are not in error in the second loop.

PeteRager committed:
Update test_init.py

PeteRager commented:
@jjlawren changes are complete, please review when you have time.  

A question you may have is why I moved the async_dispatcher_send after the ping.   What I found is when I unplugged a speaker that message would get sent and that was keeping the speaker alive beyond the ~270 second timeout so it would not get marked unavailable.   Moving it after a successful ping eliminates that problem at which point that message makes the speaker become available.

PeteRager committed:
Merge branch 'dev' into sonos-timeout-error

PeteRager committed:
Merge issues

PeteRager committed:
Add helper function to get hostname

PeteRager committed:
Fix doc string

jjlawren commented on homeassistant/components/sonos/__init__.py:
```diff
@@ -335,19 +335,30 @@ def _add_speaker(
-        for host in self.hosts:
-            ip_addr = await self.hass.async_add_executor_job(socket.gethostbyname, host)
+
+        # Loop through each configured host and verify that Soco attributes are available for it.
+        for host in self.hosts.copy():
```
This is a change in behavior. The original intention here was to obtain the visible zones from _any_ host and then break. 

That being said, I'm not sure that attempting to detect if any hosts have become unavailable in this manner will be reliable with this method. The call is made to `sync_get_visible_zones` here: https://github.com/home-assistant/core/blob/95d16c98293972e3474f27182ac1bd3e50deab7a/homeassistant/components/sonos/helpers.py#L122-L126

The first two attributes return cached values (no I/O attempted) and the last is aggressively cached from the ZoneGroupState, which is shared across all speakers in the system. This means it's very likely that only the first speaker in a loop will actually be tested (and possibly not even that one!).

If you want to actively ensure all speakers are still available, you'd need to make a non-caching call to each device, such as with this method: https://github.com/home-assistant/core/blob/95d16c98293972e3474f27182ac1bd3e50deab7a/homeassistant/components/sonos/speaker.py#L573-L578

PeteRager replied:
On startup with this scenario, the function was running into trouble.  Assume 3 hosts  listed A,B,C; A+C are on and B is off.  The first call to get visible zones for A succeeds and returns A+C, this causes the top loop to exit.

In the bottom loop, it will loop through A,B,C; when it hits B it determines it is not a known speaker and attempts to send a discovery message, which attempt to access soco.uid which is not known or cached and hence performs a blocking network call which generates an exception related to blocking the event loop which was unhandled and this would then cause setup to fail.

What I think we should try to ensure in the top loop is that we have the info cached. Such that we can exclude hosts in the error list to avoid a) logging lots of messages (prior pull request) and b) avoid event loop blocking calls in the lower loop.

When A,B,C are on, yes we will loop three times at the top, but as you state, all that information comes from cache so it's not a large perf hit.

Certainly the comment could be better.

jjlawren reviewed (commented):
Thanks for testing thoroughly. Getting the failure scenarios for Sonos devices is tricky as network layouts and behaviors vary wildly. This manual polling code should only be necessary when the various discovery methods are not operational. I have tested many similar scenarios by forcibly breaking discovery on my network but it's not a setup I enjoy re-testing.

jjlawren reviewed (approved):
I can see how this helps during startup. Should be a good improvement. Thanks again for testing and the PR 👍

PeteRager commented:
Thank you for the review. While I don't expect any issues as we now have much better unit tests. If for some reason an issue is reported on this piece of code, tag me and I'll get right on it.

bdraco commented on tests/components/sonos/test_init.py:
```diff
@@ -132,3 +140,249 @@ async def test_async_poll_manual_hosts_warnings(
+        # Test 1 first device fails, second device successful, speakers do not exist
+        with patch.object(
+            manager, "_async_handle_discovery_message", new=AsyncMock()
+        ) as mock_discovery_message, patch(
+            "homeassistant.components.sonos.async_call_later"
+        ) as mock_async_call_later, patch(
+            "homeassistant.components.sonos.async_dispatcher_send"
+        ), patch.object(
+            hass,
+            "async_add_executor_job",
+            new=AsyncMock(),
+        ) as mock_async_add_executor_job:
```
We shouldn't patch hass internals in integration tests

PeteRager replied:
Ok I'll rework tests and/or make helper functions in the class for these calls.  It'll improve readability of the test also.

PeteRager replied:
@bdraco - updated the test to remove patching of the hass object, please review.

PeteRager committed:
Remove patching of hass system object

PeteRager committed:
Update doc strings

bdraco commented on tests/components/sonos/test_init.py:
```diff
@@ -132,3 +140,254 @@ async def test_async_poll_manual_hosts_warnings(
+        # Test 2 first device successful, second device fails, speakers do not exist
+        with patch.object(
+            manager, "_async_handle_discovery_message", new=AsyncMock()
+        ) as mock_discovery_message, patch(
+            "homeassistant.components.sonos.async_call_later"
+        ) as mock_async_call_later, patch(
+            "homeassistant.components.sonos.async_dispatcher_send"
+        ), patch(
+            "homeassistant.components.sonos.sync_get_visible_zones"
+        ) as mock_sync_get_visible_zones:
```
This is better but we should not patch the following:

https://developers.home-assistant.io/docs/development_testing?_highlight=writing#writing-tests-for-integrations

> Make sure to not interact with any integration details in tests of integrations. Following this pattern will make the tests more robust for changes in the integration.


`_async_handle_discovery_message`: this is integration code 
`async_call_later`: this is considered hass internals (events)
`async_dispatcher_send`: this is considered hass internals (events)
`sync_get_visible_zones`: this is integration code. Patch the pypi code instead (ex soco.household_id, soco.ui, soco.visible_zones)

PeteRager replied:
I'm on day 4 of hitting dead ends trying to do this and could use some help from someone with a better mastery of pytest fixtures than me. 

Main issue I'm having is current test fixture causes SoCo() to return the same singleton object with a fixed ip address - regardless of the IP address provided to the constructor.  For this test we need two independent SoCo objects with different IP addresses and the ability to mock their return values in the test code to hit the various scenarios.


https://github.com/home-assistant/core/blob/21d887dd0453d0cbbf3be42f4b3650ab92fc038d/tests/components/sonos/conftest.py#L87

bdraco replied:
You can patch the patch target to be an iterable

If side_effect is an iterable then each call to the mock will return the next value from the iterable.

https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect

PeteRager replied:
Thanks, i ended up patching in a factory w/ cache so that when a soco object is being created 
```
soco = SoCo(ip_addr)
```
It'll return one whose ip address is what is specified in the call.  In test setup these can get pushed into the factory cache so the side effects for an ip_address are clear. For example if 10.10.10.1 is offline and 10.10.10.2 is online, then a speaker should get generated for 10.10.10.2 and an error message should be logged for 10.10.10.1.

PeteRager replied:
@bdraco - Success! The unit tests are definitely alot better and more concise using your recommended technique of mocking Soco.  Thanks for the feedback.  Please review when you have time.

PeteRager committed:
Use SoCo mocking for unit tests

PeteRager committed:
Updates

PeteRager committed:
Add delay

PeteRager committed:
Fix tests errors caused by merging issue

PeteRager committed:
Update conftest.py

PeteRager committed:
Move subscription timeout patch to conftest

bdraco commented on tests/components/sonos/test_init.py:
```diff
@@ -132,3 +141,260 @@ async def test_async_poll_manual_hosts_warnings(
+        {"sonos": {"media_player": {"interface_addr": "127.0.0.1"}}},
+    )
+    await hass.async_block_till_done()
+    manager: SonosDiscoveryManager = hass.data[DATA_SONOS_DISCOVERY_MANAGER]
```
I see you spent a lot of time on this so I hate to give you this feedback, but we aren't allowed to access `hass.data` in tests as it falls under integration internals. 

The idea behind the test design is it should only be patching the PyPI library code (some exceptions for patching async_setup_entry and constants) so that when we change the integration we are still testing that the whole integration remains functional.

PeteRager replied:
The existing tests in this component do this.  It's done in test_config_flow.py and test_switch.py.  I don't see another way to get the instance of the SonosDiscoveryManager.  The function under test is a method of that class.

Can we do this as a follow up pull request to fix it across all the tests in this component?

![image](https://user-images.githubusercontent.com/76050312/235494407-e618c792-8dbd-47bc-b6f1-23f1c7c32c5c.png)

bdraco replied:
An exception would need to be granted for that.  Historically that generally doesn’t happen. I will ask for a second opinion though just in case.

PeteRager replied:
Sounds good.  The existing function had no tests and had a reported issue and several latent issues contained within it. In fixing those issues, I've tried to retire the existing technical debt for the function by implementing tests that cover all branches of execution so we can verify if the function works as designed.  That work is done and is a distinct improvement over having no tests for this function.

For reference there are 685 locations in tests/ where this is done across HASS and hence this looks like a common practice.

bdraco replied:
> For reference there are 685 locations in tests/ where this is done across HASS and hence this looks like a common practice.

We have been slowly refactoring to remove them. It will take many years before they are all done at the current rate.

bdraco replied:
looks good to me

PeteRager replied:
@bdraco - i've updated all the tests to not access hass.data.  Please review when you have 5.  I did not think it was possible, but yet, I've done it.  Thanks for pushing.

bdraco replied:
nice!

bdraco commented:
Second opinion request is for https://github.com/home-assistant/core/pull/90648#discussion_r1181723342

PeteRager committed:
remove test access to hass.data

PeteRager committed:
Update test_init.py

bdraco commented on tests/components/sonos/test_init.py:
```diff
@@ -132,3 +143,254 @@ async def test_async_poll_manual_hosts_warnings(
+        with caplog.at_level(logging.DEBUG):
+            caplog.clear()
+            # Delay for second iteration and other tasks
+            await asyncio.sleep(1.0)
```
I'm assuming the sleep is waiting for something happening in another thread.

You could change this to an asyncio.Event or asyncio.Future and do

`await event.wait()`

Than mock the threaded code to do `loop.call_soon_threadsafe(event.set)` so that it unblocks here

PeteRager replied:
Yes, there a whole set of tasks that run to create the speaker and managing the subscription failing on the first run of the manual discovery code, those need to complete before the second run of the manual discovery code, then after the second run of the manual discovery code (which occurs 0.5 seconds later), we need to wait for those tasks to also run.  Since there are two ip addresses under test (which was the buggy area); multiple of those need to run.

That test is just trying to make sure that this occurs.

https://github.com/PeteRager/core/blob/f08158ec6db4196e7cdd056c259fe74f911a1da1/homeassistant/components/sonos/__init__.py#L420

```
                    async_dispatcher_send(
                        self.hass,
                        f"{SONOS_SPEAKER_ACTIVITY}-{soco.uid}",
                        "manual zone scan",
                    )
```

But that destination in speaker.py will not call anything in pypi (SoCo) because the speaker has not "cooled down" enough.

Perhaps I'll try to subscribe to that topic directly and have the callback fire the event which will then release the test code?

bdraco replied:
I think it would be better to subscribe to the dispatcher than the sleep.

The dispatcher api isn't likely to change in the future so that should be a quite safe.

PeteRager replied:
done

PeteRager committed:
Merge branch 'dev' into sonos-timeout-error

PeteRager committed:
Fix regression error, remove unneeded change

PeteRager committed:
Wait on events rather than using a sleep

bdraco commented on tests/components/sonos/test_init.py:
```diff
@@ -335,11 +370,14 @@ async def test_async_poll_manual_hosts_6(
-            await asyncio.sleep(1.0)
+            # The discovery events should not fire, wait with a timeout.
+            with pytest.raises(asyncio.TimeoutError):
+                await asyncio.wait_for(speaker_1_activity.event.wait(), 1.0)
```
`asyncio.wait_for` has some bad cancellation semantics up until cpyton 3.12 so we try not to use it.

Use `async_timeout` for now, we can replace it with `asyncio.timeout` when we drop cpython 3.10 support.


or you can do
```python
if sys.version_info[:2] < (3, 11):
    from async_timeout import timeout as asyncio_timeout
else:
    from asyncio import timeout as asyncio_timeout
```

PeteRager replied:
Updated that.  Also fixed something else that was failing in 3.11.  The best news of the day, is this last set of work revealed a bug in the code - which is now corrected.

PeteRager committed:
Updates

Don't use event wait for
Simplify waiting on multiple events

bdraco commented on tests/components/sonos/test_init.py:
```diff
@@ -84,7 +99,9 @@ async def test_async_poll_manual_hosts_warnings(
         manager, "_async_handle_discovery_message"
-    ), patch("homeassistant.components.sonos.async_call_later"), patch(
+    ), patch(
+        "homeassistant.components.sonos.async_call_later"
```
We probably shouldn’t be patching async_call_later here but that’s an existing issue

PeteRager replied:
Agreed, that's from a prior test.  I'll get that fixed in a new pull request.

PeteRager commented:
@jjlawren please review when you have 5.

jjlawren reviewed (approved):
This codepath should only be needed when network discovery is failing and manually configured hosts are required. Tests are a bit more complex now, but the previous limitation of a single mocked device is good to relax. Thanks for the thorough testing here 👍

bdraco commented:
Pushed to my production to verify that it didn't cause any side effects.  All good 👍

bdraco committed:
Merge branch 'dev' into sonos-timeout-error

bdraco commented:
Thanks @PeteRager
